src/transformers/models/musicgen/processing_musicgen.py

```diff
@@ -0,0 +1,128 @@
+
+    Args:
+        feature_extractor (`EncodecFeatureExtractor`):
+            An instance of [`EncodecFeatureExtractor`]. The feature extractor is a required input.
+        tokenizer (`T5Tokenizer`):
+            An instance of [`T5Tokenizer`]. The tokenizer is a required input.
+    """
+    feature_extractor_class = "EncodecFeatureExtractor"
+    tokenizer_class = ("T5Tokenizer", "T5TokenizerFast")
+
+    def __init__(self, feature_extractor, tokenizer):
+        super().__init__(feature_extractor, tokenizer)
+        self.current_processor = self.feature_extractor
+        self._in_target_context_manager = False
+
+    def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+        return self.tokenizer.get_decoder_prompt_ids(task=task, language=language, no_timestamps=no_timestamps)
+
+    def __call__(self, *args, **kwargs):
+        """
+        Forwards the `audio` argument to EncodecFeatureExtractor's [`~EncodecFeatureExtractor.__call__`] and the `text`
+        argument to [`~T5Tokenizer.__call__`]. Please refer to the doctsring of the above two methods for more
+        information.
+        """
+        # For backward compatibility
+        if self._in_target_context_manager:
+            return self.current_processor(*args, **kwargs)
+
+        audio = kwargs.pop("audio", None)
+        sampling_rate = kwargs.pop("sampling_rate", None)
+        text = kwargs.pop("text", None)
+        if len(args) > 0:
+            audio = args[0]
+            args = args[1:]
+
+        if audio is None and text is None:
+            raise ValueError("You need to specify either an `audio` or `text` input to process.")
+
+        if text is not None:
+            inputs = self.tokenizer(text, **kwargs)
+
+        if audio is not None:
+            audio_inputs = self.feature_extractor(audio, *args, sampling_rate=sampling_rate, **kwargs)
+
+        if audio is None:
+            return inputs
+
+        elif text is None:
+            return audio_inputs
+
+        else:
+            inputs["input_values"] = audio_inputs["input_values"]
+            if "padding_mask" in audio_inputs:
+                inputs["padding_mask"] = audio_inputs["padding_mask"]
+            return inputs
+
+    def batch_decode(self, *args, **kwargs):
+        """
+        This method forwards all its arguments to T5Tokenizer's [`~PreTrainedTokenizer.batch_decode`]. Please refer to
+        the docstring of this method for more information.
+        """
+        return self.tokenizer.batch_decode(*args, **kwargs)
+
+    def decode(self, *args, **kwargs):
+        """
+        This method forwards all its arguments to T5Tokenizer's [`~PreTrainedTokenizer.decode`]. Please refer to the
+        docstring of this method for more information.
+        """
+        return self.tokenizer.decode(*args, **kwargs)
+
+    def decode_audio(self, audio_values, padding_mask: Optional = None) -> List[np.ndarray]:
```

Comment:
FYI @sgugger this is the only new functionality since your review (the rest was just getting the doc tests to pass by fixing bugs in the docstrings)

As discussed [offline](https://huggingface.slack.com/archives/C05B51SD2RM/p1687792963817049), we want a method in the `processor` to strip any padding from our generated audio values. This method does exactly that, and is tested for with a new processor test.

The API looks as follows:
```python
from transformers import AutoProcessor, MusicgenForConditionalGeneration
from datasets import load_dataset

processor = AutoProcessor.from_pretrained("facebook/musicgen-small")
model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-small")

dataset = load_dataset("sanchit-gandhi/gtzan", split="train", streaming=True)
sample = next(iter(dataset))["audio"]

inputs = processor(
    audio=sample["audio"],
    sampling_rate=sample["sampling_rate"],
    text="80s blues track with groovy saxophone",
    padding=True,
    return_tensors="pt",
)

audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)

# NEW: post-process to remove padding from the batched audio
audio_values = processor.decode_audio(audio_values, padding_mask=inputs.padding_mask)
```

Reply:
Can we do `decode(text=...)` or `decode(audio=...)` instead? This would be more in-line with how processor deal with multimodal inputs/outputs and we have something similar in call and pad already.

Reply:
Resolved in https://github.com/huggingface/transformers/pull/24109/commits/c0235d3257e9c07281347e6ab349c831b03fa7a6 - we now decode the audios if audio values are passed, or decode token ids with the tokenizer otherwise

I had originally used the `.pad` function in the feature extractor, but this actually proved to be more complicated than a hand-written pad. This is because we want to pack the padding mask to max length with the **non-padding token**: 

1. The padding mask is constructed based on the input ids, padding left based on the length of the inputs
2. We then generate our input ids to a new max length, generating new tokens on the right always
3. The newly generated ids are all valid, so the padding mask **should not** pad these out, hence we pad with the non-padding token

This means that we have to flip the attributes of the feature extractor, perform the padding, then flip them back again
1. Flip the padding token
2. Ensure the padding side is set to `"right"`
3. Do the padding
4. Flip the padding token back again
5. Ensure the padding side is set to the original

=> overall this to me was more complicated than two lines of explicit padding